### PR TITLE
Add support for tracking crashes

### PR DESCRIPTION
**Pull request by jkmassel**
This PR adds support for using Tracks as a wrapper around any crash logging service – in this case, Sentry.

If, in the future, apps needed to work with another crash logging provider, it'd be simple enough to swap out the implementation without disrupting any of the other applications.

Additionally, a lot of the low-level bookkeeping work is now represented here, ensuring that we don't have code duplication in many different applications. This will become more important as features such as log processing are added.

**Sample implementations:**

https://github.com/Automattic/simplenote-ios/pull/304
https://github.com/Automattic/simplenote-macos/pull/325

### CODE REVIEW

**Comment by jtreanor** [comment]
Nice work @jkmassel!

Would it be possible to keep the Sentry specific logic out of Tracks? I really like the way you have made the behavior generic but I don't love the fact that `Sentry` is still a direct dependency here. What do you think?

**Comment by astralbodies** [comment]
Does this mean that Tracks will receive crash reports or that we're letting the Tracks client set up crash reporting?

**Comment by yaelirub** [comment]
Should this PR include an update to the README file?

**Comment by jkmassel** [comment]
>Does this mean that Tracks will receive crash reports or that we're letting the Tracks client set up crash reporting?

We'll be letting the Tracks client set up crash reporting – Sentry will receive crashes for the foreseeable future.

>Should this PR include an update to the README file?

Good call – will do.

**Comment by jkmassel** [comment]
@shiki – requesting your eyes on this because you had some great comments on the Android architecture.

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import XCTest
+@testable import AutomatticTracks
+
+class CrashLoggingTests: XCTestCase {
```
Comment (shiki) [inline Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: 👏 Thank you for adding tests!

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+        CrashLogging.logMessage("This is a test")
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+//    func testThatLoggedErrorsContainAStackTrace() {
```
Comment (shiki) [inline Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: ❓ What is the plan for these commented out tests?

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+//
+//        CrashLogging.start(withDataProvider: mockDataProvider)
+//        CrashLogging.logMessage("This is a test")
+//
+//        wait(for: [expectation], timeout: 1)
+//    }
+}
+
+/// Allow throwing Strings as error
+extension String: Error {}
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: It's used as a shortcut here: https://github.com/Automattic/Automattic-Tracks-iOS/pull/103/files/8d80f1f346eb45f73f7f8081284c279cee2a5933#diff-54236c2e7c3ce4c8904239c3cdc9100fR94 to avoid creating an entire `Error` object to test with! 😄

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+extension CrashLoggingTests {
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Nice! Simplified in https://github.com/Automattic/Automattic-Tracks-iOS/pull/103/commits/257f7b2917ba22ab558af8ff1bde8d54ff523854

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    fileprivate var validDSN: String { return "https://0000000000000000000000000000000@sentry.io/0000000" }
+    fileprivate var invalidDSN: String { return "foo" }
+    fileprivate var testUser: TracksUser { return TracksUser(userID: "foo", email: "bar", username: "baz") }
+}
+
+struct MockCrashLoggingDataProvider : CrashLoggingDataProvider {
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Yep, that's nice and consistent! 

https://github.com/Automattic/Automattic-Tracks-iOS/pull/103/commits/257f7b2917ba22ab558af8ff1bde8d54ff523854

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    var sentryDSN: String = ""
+    var userHasOptedOut: Bool = false
+    var buildType: String = "test"
+
+    var _currentUser: TracksUser? = nil
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Ah yes – glad we're cleaning that up :) https://github.com/Automattic/Automattic-Tracks-iOS/pull/103/commits/6a57e78574a1a93d09814573eddc81d99be81da8

File: Automattic-Tracks-iOS/CrashLogging.swift [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            updateCurrentUser()
+
+        } catch let error {
+            logError(error)
+        }
+    }
+
+    /// A Sentry hook used to attach any additional data to the event.
+    private func beforeSerializeEvent(_ event: Event) {
+        event.tags?["locale"] = NSLocale.current.languageCode
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: I wish they'd just put an `@nonnull` attribute on that one! 🙃 

It's always set by the Sentry framework in `beforeSerializeEvent`'s calling method: https://github.com/getsentry/sentry-cocoa/blob/6f1524afd64433c0230932d3b78b52570c362011/Sources/Sentry/SentryClient.m#L211-L221.

AFAIK there's no way for it to be `nil`, and if it were, the event would be so useless I'm not sure adding the locale would do much for us! 😅 

tl;dr – I think it's ok – we could be extra-defensive but it doesn't really get us anything 🤷‍♂

File: Automattic-Tracks-iOS/CrashLogging.swift [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        let result = !CrashLogging.userHasOptedOut
+        #endif
+
+        shouldSendEventCallback?(result)
+
+        return result
+    }
+
+    /// The current state of the user's choice to opt out of data collection. Provided by the data source.
+    public static var userHasOptedOut: Bool {
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: ❓Thanks for bringing this one up! So, by making this `private`, you're thinking it's better if the application developer reads this from their own `CrashLoggingDataProvider` implementation (or whatever that's calling underneath) rather than off this global class?

I can definitely be convinced of this – I left it `public` as a handy global lookup, but I think I like the simplicity of closing it off, especially since, as you pointed out, in a well-architected app it shouldn't be needed.

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/CrashLogging.swift]: Ah okay. I only looked at the `tags` declaration which is defined as `_Nullable`. 

```objc
@property(nonatomic, strong) NSDictionary<NSString *, NSString *> *_Nullable tags;
```

The [`setSharedProperties`](https://github.com/getsentry/sentry-cocoa/blob/6f1524afd64433c0230932d3b78b52570c362011/Sources/Sentry/SentryClient.m#L328) seems to set the tags only if `self.tags` is not `nil`. 

```objc
    if (nil != self.tags) {
        if (nil == event.tags) {
            event.tags = self.tags;
        } else {
            NSMutableDictionary *newTags = [NSMutableDictionary new];
            [newTags addEntriesFromDictionary:self.tags];
            [newTags addEntriesFromDictionary:event.tags];
            event.tags = newTags;
        }
    }
```

Well, if you've tested this then I guess it's fine. It's not gonna cause a crash if we're wrong. 😄

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/CrashLogging.swift]: >  you're thinking it's better if the application developer reads this from their own CrashLoggingDataProvider implementation (or whatever that's calling underneath) rather than off this global class?

Somewhat. I believe it's **not** the Crash Logger's responsibility to track this setting. It should be somewhere in the specific app's `UserPreferences` class or whatever they call it. 

`CrashLogging` should probably even call this property as `loggingEnabled: Bool` and doesn't care that it's `false` because the user opted out.

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: The problem with this is the effect is apparently global. All `String` objects will now conform to `Error`, not just in the test. I'm not sure if we want that.

File: Automattic-Tracks-iOS/TestHelpers.swift [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+typealias EventLoggingCallback = (Event) -> ()
+typealias DecisionCallback = (Bool) -> ()
+
+private var errorLoggingCallback: EventLoggingCallback? = nil
+private var messageLoggingCallback: EventLoggingCallback? = nil
+private var eventSendCallback: DecisionCallback? = nil
+
+internal extension CrashLoggingDataProvider {
+
+    var didLogErrorCallback: EventLoggingCallback? {
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/TestHelpers.swift]: I'm not completely happy with this design, tbh. I think I'd like to rework it with a `CrashLoggingDelegate` protocol, but I don't want to expose Sentry the `Event`, since that's an implementation detail. 

For now, because it's really only used as a hook system for testing (and because it's an `internal` extension, the compiler will be able to prove this is dead code and leave it out of builds) I'm ok to leave it, but I think as we add more tests, this'll need a little bit more polish :)

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift [hunk 10]
```diff
@@ -0,0 +1,141 @@
+import XCTest
+@testable import AutomatticTracks
+
+class CrashLoggingTests: XCTestCase {
+
+    private let crashLogging = CrashLogging()
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: It is indeed! After talking to @koke, we'll want to avoid using the global Sentry singleton inside of this class. I started down that road, realized it could be done later without changing the public API, and figured it'd be good to leave for later :)

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 2 above)
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: I'm not 100% sure why these fail – I'd like to reach out to Sentry to figure that out, but these are very much nice-to-haves, rather than must-haves. I suspect it's something to do with the release configuration and/or they don't run on the simulator, but I'd like to get them working at some point for full code coverage :)

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Fixed in a767cf46e0b2c19299e77fce69296774dfe18964.

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: Privatized in 65186d190954e93f2254aef8a174f169b3dd7d16!

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Yeah that's much prettier! Adjusted in 257f7b2917ba22ab558af8ff1bde8d54ff523854.

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: `self.tags` is initialized with the class, over in https://github.com/getsentry/sentry-cocoa/blob/6f1524afd64433c0230932d3b78b52570c362011/Sources/Sentry/SentryClient.m#L105!

And yeah, if all else fails, it's low-cost! 😂

File: Automattic-Tracks-iOS/TestHelpers.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/TestHelpers.swift]: > I'm ok to leave it, but I think as we add more tests, this'll need a little bit more polish :)

Sounds good to me. 👍

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 2 above)
Comment (shiki) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Cool. Perhaps a comment in the code would be nice so someone will not delete them. I know I would. 😄

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: I agree that that would be bad! But it seems to be confined to the test target?

Unless I'm missing something, which is 100% possible! :) 

Screen Shot 2019-05-29 at 9 49 55 AM

[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Oh, nice. I was wrong. 👍

File: Automattic-Tracks-iOSTests/CrashLoggingTests.swift (hunk 2 above)
Comment (jkmassel) [reply Automattic-Tracks-iOSTests/CrashLoggingTests.swift]: Agreed – added in 3f8da24a122477cc98791574e41e000e35dcfb6e

File: Automattic-Tracks-iOS/CrashLogging.swift [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     Initializes the crash logging system.
+
+     - Parameters:
+     - dataProvider: An object that will provide any required data to the crash logging system.
+
+     - SeeAlso: CrashLoggingDataProvider
+     */
+    public static func start(withDataProvider dataProvider: CrashLoggingDataProvider) {
+
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: Added thread-safe single initialization in af432ad276605b8a4c98cd78fc67e35dd4580623!

File: Automattic-Tracks-iOS/CrashLogging.swift [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+            Client.shared?.beforeSerializeEvent = sharedInstance.beforeSerializeEvent
+            Client.shared?.shouldSendEvent = sharedInstance.shouldSendEvent
+
+            // Add additional data
+            Client.shared?.releaseName = dataProvider.releaseName
+            Client.shared?.environment = dataProvider.buildType
+            updateCurrentUser()
+
+        } catch let error {
+            logError(error)
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: The only likely exception to come out if this is an invalid `DSN`. If that happens, IMHO I'd rather not crash, and just have this subsystem not work. 

But I could be convinced otherwise :)

File: Automattic-Tracks-iOS/CrashLogging.swift [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+        return user.sentryUser(withData: data)
+    }
+
+    /// Causes the Crash Logging System to refresh its knowledge about the current user.
+    ///
+    /// This is required in situations like login / logout, when the system otherwise might not
+    /// know a change has occured.
+    ///
+    /// Calling this method in these situations prevents
+    public static func updateCurrentUser() {
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: Your point is 💯, but I'm not sure about the name 🤔 

The data provider hasn't necessarily _changed_, because ideally it'd just call out to other classes without storing much (if any) internal state.

`setDataProviderHasNewDataAvailable` covers that, but is...verbose. We could do `setNewDataAvailable` but then it might be less clear to the reader what'll happen when they call it? (the inline docs could help with that...?)

Also, the `set...` syntax seems _slightly_ more at home on iOS (I'm thinking of things like `setNeedsLayout`, as a for-instance), whereas `notify` might not be as common?

There's a really good chance I'm overthinking this naming thing 🤔😅

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/CrashLogging.swift]: I just wanted to point out that no logging might actually happen. But it's fine to wait and see how this goes. 😄

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: Follow-up to this, `setNeedsDataRefresh` seems like a nice framework-integrated name.

Sorry, I'll stop 🤓-ing out about names now.

[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/CrashLogging.swift]: Yeah. You know what? Let's leave it as is for now until we find a better name. 😄

File: Automattic-Tracks-iOS/CrashLogging.swift [hunk 14]
```diff
@@ -0,0 +1,156 @@
+import Foundation
+import Sentry
+
+/// A class that provides support for logging crashes. Not compatible with Objective-C.
+public class CrashLogging {
+
+    /// A singleton is maintained, but the host application needn't be aware of its existence.
```
[the remark this answers is not in the document]
Comment (jkmassel) [reply Automattic-Tracks-iOS/CrashLogging.swift]: I might leave that for now, just because when we start adding testing, I might need to access the `internal` initializer! 

At least we don't have to worry about consumers calling it and using it wrong :)

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 13 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/CrashLogging.swift]: @jkmassel Oh, I did not see the latest `setNeedsDataRefresh`! Let's do that. That's a great name. 👍

File: Automattic-Tracks-iOS/CrashLogging.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (shiki) [reply Automattic-Tracks-iOS/CrashLogging.swift]: 👍
